README
README for uif
==============


DOCUMENTATION

The uif project has been developed for a diskless router system and provides
a mechanism to create and simplify packet filter rules. It forces you to
provide names for every value you use in order to make firewalls less
confusing.

Please have a look at the man pages for uif(8) and uif.conf(5). There
are also example configurations in the docs/ directory.

There is some LDAP support built-in, with that you can handle a big farm
of diskles router configurations. Use uif(8) and information provided in
the doc/ directory to configure the firewall fitting your needs.


BUGS/WISHLIST

uif is on Github. If you've found a bug, or have suggestions for future
versions please report it via the project's issue tracker:
https://github.com/cajus/uif/issues

If you have installed uif on Debian, you can also use the Debian BTS for
reporting bugs. As the Debian maintainer of uif is a member of the uif
upstream development team, the Debian bugs will also reach upstream quickly.


Have fun,
-Jörg Platte, Cajus Pollmeier, Mike Gabriel, Alex Owen


ChangeLog
2022-04-19 Mike Gabriel

        * release 1.99.0 (HEAD -> master, tag: 1.99.0)
        * uif.8: Update uif man page (esp. for nft support, but also convert
          from nroff to groff). (c52a64c)
        * uif.conf.5: Update uif.conf man page (esp. for nft support, but
          also convert from nroff to groff). (fa7ba69)
        * Makefile: Assure presence of new config dirs (for dropping
          snippets). (574181c)
        * uif.pl: With nft backend, fix port based rules. (4d74c0b)
        * Support file globbing and introcuce conf.d directories for dropping
          config snippets. (96ed4c7)
        * Update copyright/license information. (bd7fe84)
        * compare-v(4|6)-iptables+nft-results.sh: Add test scripts that
          compare iptables-nft and nft results. (f76c2a8)
        * uif.pl: With nft backend, resolve remaining FIXMEs. This complete
          the nft backend support. (222559a)
        * uif.pl: With nft backend, drop 'meta l4proto ipv6-icmp' prefix
          before type-wise ICMPv6 packet accepts. (945c6e0)
        * Revert "uif.pl: With nft backend, use proper counter expressions."
          (f52b023)

2022-04-18 Mike Gabriel

        * uif.pl: With nft backend, use pre-defined priority keywords instead
          of hard-coded priority numbers. (7100eb4)
        * uif.pl: With nft backend, use single ports only if ranges start and
          end with the same port. (ff27dfe)
        * uif.pl: Use camel-case key for 'FilterCommand' in $Sysconfig hash.
          (26b0453)
        * uif.pl: With nft backend, use proper counter expressions. (123c3c3)

2022-04-17 Mike Gabriel

        * uif.pl: Introduce line numbering in print mode, add cmdline option
          '-l'. (97f5523)

2022-04-16 Mike Gabriel

        * uif.pl: Convert another "-j" iptables option into "counter jump".
          (8b830a7)

2019-12-07 Mike Gabriel

        * uif.pl: Fix snat/dnat multiport rules in nft. (6fcfd7a)

2019-12-06 Mike Gabriel

        * uif.pl: Fix icmpv6 rules for nftables. (7874781)
        * uif.pl: Resolve one FIXME: snat/dnat rules for native nft command.
          (c3edde5)
        * uif.pl: Drop not required (and flawed) <FIXME> (8a85e2e)
        * mark TODOs with FIXME tags (aae4ab1)
        * Turn genRuleDump_NFT so that now generates a native nft command
          set. (a1b95d1)

2019-12-05 Mike Gabriel

        * uif.pl: Introduce FILTER_COMMAND environment parameter and support
          nft command sets as well as iptables* command sets.
          (d598df0)

2019-12-06 Mike Gabriel

        * Send "Skipping..." messages to STDERR. (98a57ec)

2019-12-05 Mike Gabriel

        * default: Parameter that shall be perceived by the uif Perl script
          need to be exported to the environment. (99d6e86)
        * uif.pl: white lines removed / added (8f2746e)
        * Prepend targets starting with ${id} by CHAIN_ (so that target names
          work in nftables, too). (d72d1d7)
        * The Linux firewall is shifting towards nftables, so let's prepare
          for that (by using the legacy iptables commands).
          (55ba357)

2018-08-20 Mike Gabriel

        * release 1.1.9 (a7ec235) (tag: 1.1.9)
        * Makefile: Fix flawed usage of DESTDIR and PREFIX. (1448fd3)

2017-01-16 Mike Gabriel

        * uif.spec: Date fix in changelog. (ee38674)

2017-01-15 Mike Gabriel

        * release 1.1.8 (1123e0c) (tag: 1.1.8)
        * validateData: Prevent from using networks with MAC addresses
          neither for outward bound nor for destination networks.
          (d913b96)
        * Better test and fix MAC address based source filtering. (477c55b)
        * release 1.1.7 (9d20e8e) (tag: 1.1.7)
        * MAC-source filtering, regression fix: Re-add push of $ip to
          @netobjects if network object contains a MAC address.
          (bd85804)
        * UIF: Allow MAC syntax in network items only with real IP addresses,
          not with DNS resolvable host names. (36d6d86)
        * release 1.1.6 (3435812) (tag: 1.1.6)
        * Makefile: Fix installation of uif.initscript. (0038831)
        * uif.conf.5: Add some hints and notes about IPv6 support. (9c01c91)
        * doc files: consistently use UIF in capital letters. (6d01761)
        * services file: Add some more services (kerberos5 et al., ldaps,
          swat, openvpn, mysql, munin, cfenging, xmpp-client,
          xmpp-server, icinga2, webmin and puppet). (868ac44)
        * COPYRIGHT: Update copyright date for Mike Gabriel. (cfee841)
        * ChangeLog: Convert to GNU ChangeLog style, generated from Git
          history. (577cc6b)
        * INSTALL.md: Typo fix in Perl module name. (d75e0ed)
        * Drop former uif initscript, replaced by new uif.initscript file.
          (8766e7b)
        * Revert "release 1.1.6" (26c5c19)
        * Rename README.IPv6 -> README.IPv6.md (465c6da)
        * release 1.1.6 (ed896f4)
        * Update most documentation files and convert to markdown syntax.
          (f2f7bf2)
        * init script: Adopting Debian's init script as an example into
          upstream code. (13adace)
        * Add VERSION file (with last released version number). (402d8f5)
        * Drop debian/ packaging folder, we are an upstream project.
          (6aae291)
        * IPv6 name resolution: Work around broken IPv6 name resolution in
          NetAddr::IP (see: CPAN issue #119858). (91d3907)

2017-01-13 Mike Gabriel

        * IPv6 support: More locations in the code spotted, where we need to
          differentiate between IPv4 and IPv6 mode. (2baab0e)

2016-04-18 Mike Gabriel

        * Merge branch 'ka7-spelling_fix' (91cb12b)

2016-04-16 klemens

        * spelling fix, as of lintian.debian.org (ae13340)

2015-03-11 Mike Gabriel

        * release 1.1.5 (01892a9) (tag: 1.1.5)
        * bump version and dates (45b9141)
        * Fix severe flaw in IPv4-only/IPv6-only rule setup. Don't open IPv4
          wholes when setting up IPv6-only rules and vice versa.
          (d8c8700)


compare-v4-iptables+nft-results.sh
#!/bin/bash

# Copyright (C) 2022 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

# this script requires /etc/uif/ to be in place and configured well

tmpresults_nft="$(mktemp)"
tmpresults_iptables="$(mktemp)"

sudo nft flush ruleset ip

sudo FILTER_COMMAND=nft ./uif.pl
sudo nft list ruleset ip 1> "${tmpresults_nft}"
sudo nft flush ruleset ip

sudo FILTER_COMMAND=iptables-nft ./uif.pl
sudo nft list ruleset ip 1> "${tmpresults_iptables}"
#sudo nft flush ruleset ip

diff -wu "${tmpresults_iptables}" "${tmpresults_nft}"


compare-v6-iptables+nft-results.sh
#!/bin/bash

# Copyright (C) 2013-2022 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

# this script requires /etc/uif/ to be in place and configured well

tmpresults_nft="$(mktemp)"
tmpresults_iptables="$(mktemp)"

sudo nft flush ruleset ip6

sudo FILTER_COMMAND=nft ./uif.pl -6
sudo nft list ruleset ip6 1> "${tmpresults_nft}"
sudo nft flush ruleset ip6

sudo FILTER_COMMAND=iptables-nft ./uif.pl -6
sudo nft list ruleset ip6 1> "${tmpresults_iptables}"
#sudo nft flush ruleset ip6

diff -wu "${tmpresults_iptables}" "${tmpresults_nft}"
